pandas/io/sas/saslib.pyx

```diff
@@ -189,49 +180,205 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
+        self.byte_chunk = parser._byte_chunk
+        self.string_chunk = parser._string_chunk
+        self.row_length = parser.row_length
+        self.cached_page = <char *>parser._cached_page
+        self.bit_offset = self.parser._page_bit_offset
+        self.subheader_pointer_length = self.parser._subheader_pointer_length
+        self.is_little_endian = parser.byte_order == "<"
+        self.column_types = np.empty(self.column_count, dtype=long)
+
+        column_types = parser.column_types
+
+        # map column types
+        for j in range(self.column_count):
+            if column_types[j] == b'd':
+                self.column_types[j] = 1
+            elif column_types[j] == b's':
+                self.column_types[j] = 2
             else:
-                byte_chunk[jb, s:s+length] = np.frombuffer(temp, dtype=np.uint8)
-            jb += 1
-        elif parser.column_types[j] == b's':
-            string_chunk[js, parser._current_row_in_chunk_index] = bytes(temp)
-            js += 1
+              raise ValueError("unknown column type: %s" % self.parser.columns[j].ctype)
+
+        # compression
+        if parser.compression == const.rle_compression:
+            self.decompress = rle_decompress
+        elif parser.compression == const.rdc_compression:
+            self.decompress = rdc_decompress
         else:
-            raise ValueError("unknown column type: %s" % parser.columns[j].ctype)
+            self.decompress = NULL
+
+    def read(self, int nrows):
+        cdef:
+            bint done
+            int i
+
+        for i in range(nrows):
+            done = self.readline()
+            if done:
+                break
+
+        # update the parser
+        self.parser._current_row_on_page_index = self.current_row_on_page_index
+        self.parser._current_row_in_chunk_index = self.current_row_in_chunk_index
+        self.parser._current_row_in_file_index = self.current_row_in_file_index
+
+    cdef bint read_next_page(self):
+        cdef done
+
+        done = self.parser._read_next_page()
+        if done:
+            self.cached_page = NULL
+        else:
+            self.cached_page = <char *>self.parser._cached_page
+            self.current_row_on_page_index = 0
+        return done
+
+    cdef bint readline(self):
+
+        cdef:
+            int offset, bit_offset, align_correction, subheader_pointer_length
+            bint done
+
+        bit_offset = self.bit_offset
+        subheader_pointer_length = self.subheader_pointer_length
+
+        # If there is no page, go to the end of the header and read a page.
+        if self.cached_page == NULL:
+            self.parser._path_or_buf.seek(self.parser.header_length)
+            done = self.read_next_page()
+            if done:
+                return True
+
+        # Loop until a data row is read
+        while True:
+            if self.parser._current_page_type == const.page_meta_type:
+                flag = (self.current_row_on_page_index >=
```

Comment:
FYI there is a fair amount more that can be typed here but would involve some rewriting